### PR TITLE
Delete series support for TSM

### LINKED ISSUE #6468: HELP: influxdb cannot start back up - 'panic: runtime error: slice bounds out of range'
### Bug report

__System info: __ [Linux: 3.10.0-327.13.1.el7.x86_64, influxdb-0.12.2-1.x86_64]

__Steps to reproduce:__

1. I ran out of space in /var/lib/ (where I had influxdb).
2. Now I cannot start influxdb.

__Expected behavior:__ [influxdb should start]

__Actual behavior:__ [Cannot start influxdb]

__Additional info:__ [Include gist of relevant config, logs, etc.]

influxdb.log states:
[tsm1] 2016/04/25 12:52:12 /d0/influxdb/data/_internal/monitor/23 database index loaded in 3.312297ms
[store] 2016/04/25 12:52:12 /d0/influxdb/data/_internal/monitor/23 opened in 15.019705488s
[tsm1] 2016/04/25 12:52:17 /d0/influxdb/data/regression/default/12 database index loaded in 20.03670118s
[store] 2016/04/25 12:52:17 /d0/influxdb/data/regression/default/12 opened in 20.087676684s
panic: runtime error: slice bounds out of range

goroutine 20 [running]:
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*WriteWALEntry).UnmarshalBinary(0xc208139a68, 0xc20d8aa000, 0xb1, 0x100000, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/wal.go:593 +0x68f
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*WALSegmentReader).Next(0xc2093560c0, 0xc20c9e1c00)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/wal.go:804 +0x7be
github.com/influxdata/influxdb/tsdb/engine/tsm1.func·001(0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/cache.go:444 +0x3a1
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*CacheLoader).Load(0xc208146320, 0xc208082e80, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/cache.go:466 +0x10f
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).reloadCache(0xc208176fd0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:663 +0xe2
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).Open(0xc208176fd0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:159 +0x1dd
github.com/influxdata/influxdb/tsdb.func·004(0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/shard.go:149 +0x26a
github.com/influxdata/influxdb/tsdb.(*Shard).Open(0xc208050fc0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/shard.go:159 +0x6f
github.com/influxdata/influxdb/tsdb.func·009(0xc208031c20, 0xc20801f952, 0xa, 0xc2080ecf1d, 0x7, 0xc2080ed165, 0x2)
        /root/go/src/github.com/influxdata/influxdb/tsdb/store.go:162 +0x65f
created by github.com/influxdata/influxdb/tsdb.(*Store).loadShards
        /root/go/src/github.com/influxdata/influxdb/tsdb/store.go:170 +0xac8

*Please note, the quickest way to fix a bug is to open a Pull Request.*

### PR DESCRIPTION
**Pull request by jwilder**
###### Required for all non-trivial PRs
- [x] Rebased/mergable
- [x] Tests pass
- [x] CHANGELOG.md updated

This PR adds the capability to the TSM engine to remove points based on time ranges to support the `DELETE SERIES` statement.  The query parser/execution is not part of this PR.

The main changes to support this capability are:

* Changed the tombstone file format to also store time ranges per key
* Update `indirectIndex` to support partially deleted keys
* Update WAL to store time ranges with deleted keys
* Update Cache to allow removing keys by ranges of time
* Update CacheLoader to handle keys that are partially deleted
* Update Compactor to skip and filter out blocks and values that are partially deleted

There are a few unrelated changes in the PR done to make the implementation simpler as well as fix some bugs:

* `directIndex` and `indirectIndex` implemented the same `TSMIndex` interface which is used for reading and writing.  We really only used `directIndex` for writing and tests and `indirectIndex` for reading.  These have been split out into separate reader and writer types so that we do not need to continue to implement reading capabilities in the `directIndex` when they are not used.  This will also make it easier to support different version of indexes in the future when we need to extend the TSM file format.
* The prior change allowed for a lot of code to be removed (mainly `fileAccessor`)
* Fixed many places where corrupted WAL segments could panic.  Should fix #6468 and similar ones in the future.
* Removed `NewTSMReaderWithOptions` which allowed creating a TSMReader w/ a direct or indirect index.  There is only `NewTSMReader` now that always uses `indirectIndex`.  Create a `TSMReader` with a `directIndex` was only useful for writing tests that don't write to disk.
* Removed several inefficient `TSMReader` funcs that should never be called (e.g. `Keys`)

cc @benbjohnson

### CODE REVIEW

File: tsdb/engine/tsm1/cache.go [hunk 1]
```diff
@@ -306,11 +324,24 @@ func (c *Cache) Values(key string) Values {
 
 // Delete will remove the keys from the cache
 func (c *Cache) Delete(keys []string) {
+	c.DeleteRange(keys, math.MinInt64, math.MaxInt64)
+}
+
+// DeleteRange will remove the keys containing points between min and max from the cache
```
Comment (joelegasse) [inline tsdb/engine/tsm1/cache.go]: This should probably be re-worded to mention that only the values within the given range are removed. Right now, I interpret this to mean the entire key is removed if any value is in that time range.

File: tsdb/engine/tsm1/cache.go [hunk 2]
```diff
@@ -306,11 +324,24 @@ func (c *Cache) Values(key string) Values {
...
 	for _, k := range keys {
-		delete(c.store, k)
+		origSize := c.store[k].size()
+		if min == math.MinInt64 && max == math.MaxInt64 {
+			c.size -= uint64(origSize)
+			delete(c.store, k)
+			continue
+		}
+
+		c.store[k].filter(min, max)
```
Comment (joelegasse) [inline tsdb/engine/tsm1/cache.go]: We should probably check to see if all  the values were removed for this key here, and then delete it if it is now empty.

File: tsdb/engine/tsm1/cache_test.go [hunk 3]
```diff
@@ -72,6 +73,84 @@ func TestCache_CacheWriteMulti(t *testing.T) {
...
+	}
+
+	if got, exp := len(c.Values("bar")), 1; got != exp {
+		t.Fatalf("cache values mismatch: got %v, exp %v", got, exp)
+	}
+
+	if got, exp := len(c.Values("foo")), 3; got != exp {
+		t.Fatalf("cache values mismatch: got %v, exp %v", got, exp)
+	}
+}
```
Comment (joelegasse) [inline tsdb/engine/tsm1/cache_test.go]: Possibly add a test case regarding the key removal above... `c.DeleteRange([]string{"foo"}, 0, 4)` then verify that `foo` is no longer a key in the cache?

File: tsdb/engine/tsm1/engine.go [hunk 4]
```diff
@@ -368,7 +369,7 @@ func (e *Engine) WritePoints(points []models.Point) error {
 }
 
 // DeleteSeries deletes the series from the engine.
```
Comment (joelegasse) [inline tsdb/engine/tsm1/engine.go]: Update the comment to reflect the time range, "DeleteSeries removes points from the series whose times fall between min and max. The entire series is removed if all points fall within the specified range."

Comment (joelegasse) [inline tsdb/engine/tsm1/engine.go]: Or... Keep `DeleteSeries` and add `DeleteSeriesRange` that takes in the time range. Similar to the changes in the cache.

File: tsdb/engine/tsm1/reader.go [hunk 5]
```diff
@@ -600,12 +627,76 @@ func (d *indirectIndex) Delete(keys []string) {
...
+}
+
 func (d *indirectIndex) Contains(key string) bool {
 	return len(d.Entries(key)) > 0
 }
 
 func (d *indirectIndex) ContainsValue(key string, timestamp int64) bool {
-	return d.Entry(key, timestamp) != nil
+	entry := d.Entry(key, timestamp)
+	if entry != nil {
```
Comment (joelegasse) [inline tsdb/engine/tsm1/reader.go]: [Coding Style] The convention in Go is to return early if possibly, then continue the method without being indented in a block. Changing this condition to `if entry == nil { return false }` would remove the large block here

**Comment by benbjohnson** [comment]
Overall it looks really good. There were a couple nits mentioned by @joelegasse but other than that, 👍 .

Also, I'm happy that I'm not the only one submitting 2KLOC PRs. :)

**Comment by jwilder** [comment]
@benbjohnson It's mostly test code. :)

All comments addressed.  I brought back `engine.DeleteSeries` and had that call `engine.DeleteSeriesRange`.

**Comment by joelegasse** [comment]
LGTM 👍

**Pull request merged**